File: src/com/easy/AddDigits_258.java
/**
 * 
 */
package com.easy;

/**
 * @author Chris Lin
 * @version 1.0
 * 
 */
public class AddDigits_258 {
    // Use recursive
    // Time Complexity: O(logn)
    // Space Complexity: O(logn)
    // Runtime: 1 ms, faster than 100.00%
    // Memory Usage: 33.8 MB, less than 5.02%
    public int addDigits(int num) {
        while (num >= 10) {
            num = num / 10 + num % 10;
        }

        return num;
    }

    // Use math formula to judge 9 modulation
    // Time Complexity: O(1)
    // Space Complexity: O(1)
    // Runtime: 1 ms, faster than 100.00%
    // Memory Usage: 33.4 MB, less than 5.02%
    public int addDigits2(int num) {
        if (num == 0) {
            return 0;
        }
        if (num % 9 == 0) {
            return 9;
        }
        return num % 9;
    }
}

File: Golang_src/easy/addDigits_258.go
package easy

// Use recursive:
// Time Complexity: O(logn)
// Space Complexity:O(1)
// Runtime: 7 ms, faster than 9.72%
// Memory Usage: 2.2 MB, less than 75.00%
func addDigits(num int) int {
	for num/10 != 0 {
		tmp := num
		num = 0
		for tmp != 0 {
			num += tmp % 10
			tmp /= 10
		}
	}

	return num
}

// Use math of digital root:
// Time Complexity: O(1)
// Space Complexity:O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2 MB, less than 90.28%
func addDigits2(num int) int {
	return (num-1)%9 + 1
}
